gitlab/auth.go

```diff
@@ -0,0 +1,276 @@
+// actions, like e.g. deleting a repository.
+func WithDestructiveAPICalls(destructiveActions bool) ClientOption {
+	return buildCommonOption(gitprovider.CommonClientOptions{EnableDestructiveAPICalls: &destructiveActions})
+}
+
+// WithPreChainTransportHook registers a ChainableRoundTripperFunc "before" the cache and authentication
+// transports in the chain. For more information, see NewClient, and gitprovider.CommonClientOptions.PreChainTransportHook.
+func WithPreChainTransportHook(preRoundTripperFunc gitprovider.ChainableRoundTripperFunc) ClientOption {
+	// Don't allow an empty value
+	if preRoundTripperFunc == nil {
+		return optionError(fmt.Errorf("preRoundTripperFunc cannot be nil: %w", gitprovider.ErrInvalidClientOptions))
+	}
+
+	return buildCommonOption(gitprovider.CommonClientOptions{PreChainTransportHook: preRoundTripperFunc})
+}
+
+// WithPostChainTransportHook registers a ChainableRoundTripperFunc "after" the cache and authentication
+// transports in the chain. For more information, see NewClient, and gitprovider.CommonClientOptions.WithPostChainTransportHook.
+func WithPostChainTransportHook(postRoundTripperFunc gitprovider.ChainableRoundTripperFunc) ClientOption {
+	// Don't allow an empty value
+	if postRoundTripperFunc == nil {
+		return optionError(fmt.Errorf("postRoundTripperFunc cannot be nil: %w", gitprovider.ErrInvalidClientOptions))
+	}
+
+	return buildCommonOption(gitprovider.CommonClientOptions{PostChainTransportHook: postRoundTripperFunc})
+}
+
+// WithOAuth2Token initializes a Client which authenticates with GitLab through an OAuth2 token.
+// oauth2Token must not be an empty string.
+func WithOAuth2Token(oauth2Token string) ClientOption {
+	// Don't allow an empty value
+	if len(oauth2Token) == 0 {
+		return optionError(fmt.Errorf("oauth2Token cannot be empty: %w", gitprovider.ErrInvalidClientOptions))
+	}
+
+	return &clientOptions{AuthTransport: oauth2Transport(oauth2Token)}
+}
+
+func oauth2Transport(oauth2Token string) gitprovider.ChainableRoundTripperFunc {
+	return func(in http.RoundTripper) http.RoundTripper {
+		// Create a TokenSource of the given access token
+		ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: oauth2Token})
+		// Create a Transport, with "in" as the underlying transport, and the given TokenSource
+		return &oauth2.Transport{
+			Base:   in,
+			Source: oauth2.ReuseTokenSource(nil, ts),
+		}
+	}
+}
+
+// makeOptions assembles a clientOptions struct from ClientOption mutator functions.
+func makeOptions(opts ...ClientOption) (*clientOptions, error) {
+	o := &clientOptions{}
+	for _, opt := range opts {
+		if err := opt.ApplyToGitlabClientOptions(o); err != nil {
+			return nil, err
+		}
+	}
+	return o, nil
+}
+
+// NewClientFromPAT creates a new gitlab.Client instance for GitLab API endpoints.
+func NewClientFromPAT(personalAccessToken string, optFns ...ClientOption) (gitprovider.Client, error) {
```

Comment:
Can we pull out the common code between these client factories easily? Can make a follow up issue too. Maybe add a FIXME as a note if viable.

Reply:
Yep, should be straightforward! need to test the OAuth token and username and password flows, and add support for self-hosted domains still.

Reply:
Merged the functions to `NewClient()` and tested with `gitlab.dev.wkp.weave.works`. The `sshDomain` field is the last related test for this I can see